### PR TITLE
Less conv_tab allocations when pushing relevances, esp skip_pattern

### PR DESCRIPTION
**Pull request by SkySkimmer**
https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/656/console
~~~
┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬────────────────────────────────┬───────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]      │        mem faults         │
│                        │                         │                                             │                                             │                                │                           │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD  PDIFF   │     NEW    OLD    PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│           coq-compcert │ 1018.83 1038.77 -1.92 % │     2837855535871     2892600277777 -1.89 % │     4405826800620     4407340336293 -0.03 % │    1291784    1291636  +0.01 % │     265    320   -17.19 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│              coq-flocq │  569.10  577.00 -1.37 % │     1582918771705     1605187659673 -1.39 % │     2296197146161     2301267132702 -0.22 % │    1729424    1545896 +11.87 % │     357    425   -16.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│        coq-fiat-crypto │ 3999.73 4015.44 -0.39 % │    11107503795691    11154835276907 -0.42 % │    17836253951766    17843302708015 -0.04 % │    2542464    2492228  +2.02 % │    1107    817   +35.50 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│       coq-fiat-parsers │  699.42  701.05 -0.23 % │     1955057728479     1961319347779 -0.32 % │     2978996895059     2980635082310 -0.05 % │    2878308    2889768  -0.40 % │     378    661   -42.81 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│             coq-sf-plf │   43.92   44.00 -0.18 % │      123450188628      123180057088 +0.22 % │      155714222580      155696812980 +0.01 % │     498996     499068  -0.01 % │      40      0     +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│  coq-mathcomp-fingroup │   55.72   55.77 -0.09 % │      155586928054      155072603886 +0.33 % │      202469709852      202428722482 +0.02 % │     582212     579652  +0.44 % │      12      2  +500.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│ coq-mathcomp-odd-order │ 1411.48 1412.68 -0.08 % │     3930593686157     3935654619116 -0.13 % │     6472529367370     6471246891143 +0.02 % │    1346760    1350328  -0.26 % │      39     66   -40.91 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│    coq-formal-topology │   30.54   30.54 +0.00 % │       84199125191       84019075117 +0.21 % │      109321342924      109341877511 -0.02 % │     470140     470188  -0.01 % │      21      2  +950.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│        coq-lambda-rust │ 2072.33 2070.44 +0.09 % │     5776748124692     5766168081625 +0.18 % │     7940632786847     7939787693826 +0.01 % │    1380220    1396372  -1.16 % │      49      0     +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│  coq-mathcomp-solvable │  195.33  195.08 +0.13 % │      545211597480      543690389257 +0.28 % │      728883869392      729474975815 -0.08 % │     841220     860700  -2.26 % │      13      2  +550.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│         coq-coquelicot │   74.92   74.80 +0.16 % │      205669426150      205587659489 +0.04 % │      247123797801      247102567701 +0.01 % │     693460     692796  +0.10 % │     214     11 +1845.45 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│         coq-verdi-raft │ 1404.92 1402.01 +0.21 % │     3920694975488     3908420104368 +0.31 % │     5316917777127     5317108002928 -0.00 % │    1811156    1811968  -0.04 % │       0     37  -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│       coq-math-classes │  220.12  219.54 +0.26 % │      614639755913      613133727340 +0.25 % │      776134938327      776160421492 -0.00 % │     517320     517000  +0.06 % │      41     19  +115.79 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│               coq-hott │  347.04  346.11 +0.27 % │      944029295309      942137597608 +0.20 % │     1422264464077     1422321250628 -0.00 % │     479616     479792  -0.04 % │     457    202  +126.24 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│              coq-color │  694.40  692.45 +0.28 % │     1941192362478     1933800660418 +0.38 % │     2311047277378     2310312905028 +0.03 % │    1392656    1379108  +0.98 % │     183    247   -25.91 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│             coq-geocoq │ 2038.79 2033.03 +0.28 % │     5687816251229     5673088395748 +0.26 % │     8205266245496     8205428867519 -0.00 % │    1199804    1199896  -0.01 % │     283    265    +6.79 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│               coq-corn │ 1515.88 1510.88 +0.33 % │     4230578813156     4216395883441 +0.34 % │     6218090460349     6218024246782 +0.00 % │     859944     867368  -0.86 % │       1      1    +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│ coq-mathcomp-character │  257.59  256.51 +0.42 % │      717527473094      714953984727 +0.36 % │      983800305729      985093132430 -0.13 % │    1055492    1059864  -0.41 % │      79    109   -27.52 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│          coq-fiat-core │  106.97  106.50 +0.44 % │      303664405781      302003302047 +0.55 % │      378293268693      377351646255 +0.25 % │     493148     492880  +0.05 % │     415    258   +60.85 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│   coq-mathcomp-algebra │  155.77  155.04 +0.47 % │      434187066239      432753309361 +0.33 % │      547065415700      547092469999 -0.00 % │     637352     637556  -0.03 % │      10     45   -77.78 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│            coq-bignums │   71.32   70.94 +0.54 % │      198226985553      197683139061 +0.28 % │      260707391591      260637263359 +0.03 % │     525244     525232  +0.00 % │     182    154   +18.18 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│     coq-mathcomp-field │  228.61  227.35 +0.55 % │      636927681850      633309256277 +0.57 % │      919780810493      919981199328 -0.02 % │     749512     750348  -0.11 % │       0      0     +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│            coq-unimath │ 2100.52 2088.76 +0.56 % │     5848219560782     5816319081172 +0.55 % │    10204563533204    10202375958251 +0.02 % │     910076     946712  -3.87 % │     528    166  +218.07 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│              coq-verdi │  126.75  125.57 +0.94 % │      352424094992      349587543288 +0.81 % │      451329426563      450331657318 +0.22 % │     614880     614796  +0.01 % │      13      4  +225.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────────────┤
│ coq-mathcomp-ssreflect │   39.74   39.32 +1.07 % │      109421265925      109118942216 +0.28 % │      130477591457      130442574458 +0.03 % │     527360     527348  +0.00 % │     134     10 +1240.00 % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴────────────────────────────────┴───────────────────────────┘
~~~

### CODE REVIEW

File: kernel/reduction.ml [hunk 1]
```diff
@@ -315,11 +315,19 @@ type conv_tab = {
 let push_relevance infos r =
   { infos with relevances = r.Context.binder_relevance :: infos.relevances }
 
-let rec skip_pattern infos n c1 c2 =
-  if Int.equal n 0 then infos, c1, c2
+let push_relevances infos nas =
+  { infos with relevances = Array.fold_left (fun l x -> x.Context.binder_relevance :: l) infos.relevances nas }
+
+let rec skip_pattern infos relevances n c1 c2 =
+  if Int.equal n 0 then {infos with relevances}, c1, c2
```
Comment (ppedrot) [inline kernel/reduction.ml]: You can actually return directly `infos` here.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@maximedenes are you fine with this?

**Comment by maximedenes** [comment]
Is the 1% increase case noise or not?

**Comment by ppedrot** [comment]
I believe so, the cycles (which are more reliable) are not matching up.

**Comment by maximedenes** [comment]
How about trying a new run, then?

**Comment by ppedrot** [comment]
@SkySkimmer ping.

**Comment by SkySkimmer** [comment]
New run https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/665/console
~~~

┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬─────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │       mem faults        │
│                        │                         │                                             │                                             │                               │                         │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │     NEW     OLD PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│           coq-compcert │ 1019.67 1038.60 -1.82 % │     2841629983026     2890220308475 -1.68 % │     4406017617212     4407581173359 -0.04 % │    1291808    1291852 -0.00 % │     422     332 +27.11 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│              coq-flocq │  568.90  572.77 -0.68 % │     1583045290640     1593213447158 -0.64 % │     2296661059696     2293388026831 +0.14 % │    1545884    1545720 +0.01 % │     258     458 -43.67 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│        coq-fiat-crypto │ 3995.88 4012.64 -0.42 % │    11103207954233    11145707815035 -0.38 % │    17835101523601    17838948301283 -0.02 % │    2465424    2541380 -2.99 % │    1112    1015 +9.56 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│       coq-fiat-parsers │  699.13  701.05 -0.27 % │     1954640567340     1960161125893 -0.28 % │     2979488779225     2980682148655 -0.04 % │    2878336    2890584 -0.42 % │     353     661 -46.60 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│ coq-mathcomp-odd-order │ 1408.51 1411.42 -0.21 % │     3924989792314     3934768973492 -0.25 % │     6470474065026     6470594564551 -0.00 % │    1348216    1350724 -0.19 % │      52     165 -68.48 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│ coq-mathcomp-ssreflect │   39.44   39.48 -0.10 % │      109617641718      109311688347 +0.28 % │      130567336686      130567778484 -0.00 % │     527264     527384 -0.02 % │      12      60 -80.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│               coq-hott │  346.72  346.95 -0.07 % │      945061298989      943673314450 +0.15 % │     1424213305976     1424133721020 +0.01 % │     479824     479648 +0.04 % │     463     252 +83.73 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│     coq-mathcomp-field │  227.86  227.93 -0.03 % │      634866654003      633951830634 +0.14 % │      919912888597      920141023569 -0.02 % │     750332     751408 -0.14 % │       0       6 -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│  coq-mathcomp-solvable │  195.26  195.16 +0.05 % │      545440561195      544211094161 +0.23 % │      729023275465      729337820060 -0.04 % │     841112     860636 -2.27 % │      10      86 -88.37 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│        coq-lambda-rust │ 2072.51 2068.71 +0.18 % │     5774701759193     5765830134947 +0.15 % │     7940328715488     7939934015971 +0.00 % │    1398768    1375932 +1.66 % │      49       0  +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│         coq-verdi-raft │ 1406.01 1403.12 +0.21 % │     3920628402856     3915277508692 +0.14 % │     5316917786653     5317081764420 -0.00 % │    1812164    1811320 +0.05 % │       0       0  +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│               coq-corn │ 1514.60 1511.27 +0.22 % │     4228424839801     4218469082660 +0.24 % │     6217545521041     6217427651233 +0.00 % │     860080     867552 -0.86 % │       1       1 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│             coq-geocoq │ 2036.30 2031.06 +0.26 % │     5683644026117     5665802676248 +0.31 % │     8206792309411     8205179527564 +0.02 % │    1199776    1224976 -2.06 % │     367     470 -21.91 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│    coq-formal-topology │   30.77   30.66 +0.36 % │       84260034527       83923602290 +0.40 % │      109472953797      109476063889 -0.00 % │     470172     470204 -0.01 % │      73       2 +3550.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│ coq-mathcomp-character │  257.17  256.21 +0.37 % │      716919716857      714564952186 +0.33 % │      983876372769      985195202248 -0.13 % │    1055564    1059860 -0.41 % │      87      60 +45.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│            coq-bignums │   71.42   71.14 +0.39 % │      198692030607      197752224469 +0.48 % │      260793033255      260745102246 +0.02 % │     525288     525176 +0.02 % │     179     149 +20.13 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│              coq-color │  694.62  691.69 +0.42 % │     1943026659560     1935104507958 +0.41 % │     2311077982259     2310367402086 +0.03 % │    1394624    1394428 +0.01 % │     359     566 -36.57 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│   coq-mathcomp-algebra │  156.12  155.46 +0.42 % │      435054287858      433528373514 +0.35 % │      547232804132      547204493984 +0.01 % │     637276     637940 -0.10 % │      16       3 +433.33 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│  coq-mathcomp-fingroup │   55.79   55.53 +0.47 % │      155838783097      155163812722 +0.44 % │      202556752550      202546184050 +0.01 % │     582276     579568 +0.47 % │      85       7 +1114.29 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│            coq-unimath │ 2099.16 2088.16 +0.53 % │     5848709398136     5818421949022 +0.52 % │    10206277248837    10206013531339 +0.00 % │     946788     946884 -0.01 % │     510     156 +226.92 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│       coq-math-classes │  219.98  218.78 +0.55 % │      614800909113      612317188529 +0.41 % │      776212947032      776265440168 -0.01 % │     517112     517136 -0.00 % │      44      19 +131.58 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│         coq-coquelicot │   75.27   74.65 +0.83 % │      206036316070      205688419084 +0.17 % │      247278889317      247125478367 +0.06 % │     693388     692956 +0.06 % │     214      11 +1845.45 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│              coq-verdi │  126.85  125.72 +0.90 % │      351866694879      349423011401 +0.70 % │      451522688446      450449569321 +0.24 % │     603848     605756 -0.31 % │      10       4 +150.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│          coq-fiat-core │  107.22  106.02 +1.13 % │      303564770798      302163299438 +0.46 % │      378386466508      377475622629 +0.24 % │     493092     492980 +0.02 % │     698     261 +167.43 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│             coq-sf-plf │   44.36   43.82 +1.23 % │      123501862041      123427936281 +0.06 % │      155827718250      155818466245 +0.01 % │     499096     499236 -0.03 % │      32       0  +nan % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴─────────────────────────┘
~~~

**Comment by ppedrot** [comment]
This looks like it was noise indeed. Can I proceed to merging this PR, @maximedenes?

**Comment by maximedenes** [comment]
> Can I proceed to merging this PR, @maximedenes?

Yes.

**Pull request merged**